flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSink.java

```diff
@@ -0,0 +1,250 @@
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.ProjectTopicName;
+import com.google.pubsub.v1.PubsubMessage;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+
+import java.io.IOException;
+
+/**
+ * A sink function that outputs to PubSub.
+ *
+ * @param <IN> type of PubSubSink messages to write
+ */
+public class PubSubSink<IN> extends RichSinkFunction<IN> {
+
+	private SerializableCredentialsProvider serializableCredentialsProvider;
+	private SerializationSchema<IN> serializationSchema;
+	private String projectName;
+	private String topicName;
+	private String hostAndPort = null;
+
+	private transient Publisher publisher;
+
+	private PubSubSink() {
+	}
+
+	void setSerializableCredentialsProvider(SerializableCredentialsProvider serializableCredentialsProvider) {
+		this.serializableCredentialsProvider = serializableCredentialsProvider;
+	}
+
+	void setSerializationSchema(SerializationSchema<IN> serializationSchema) {
+		this.serializationSchema = serializationSchema;
+	}
+
+	void setProjectName(String projectName) {
+		this.projectName = projectName;
+	}
+
+	void setTopicName(String topicName) {
+		this.topicName = topicName;
+	}
+
+	/**
+	 * Set the custom hostname/port combination of PubSub.
+	 * The ONLY reason to use this is during tests with the emulator provided by Google.
+	 *
+	 * @param hostAndPort The combination of hostname and port to connect to ("hostname:1234")
+	 */
+	void withHostAndPort(String hostAndPort) {
+		this.hostAndPort = hostAndPort;
+	}
+
+	void initialize() throws IOException {
+		if (serializableCredentialsProvider == null) {
+			serializableCredentialsProvider = SerializableCredentialsProvider.credentialsProviderFromEnvironmentVariables();
+		}
+		if (serializationSchema == null) {
+			throw new IllegalArgumentException("The serializationSchema has not been specified.");
+		}
+		if (projectName == null) {
+			throw new IllegalArgumentException("The projectName has not been specified.");
+		}
+		if (topicName == null) {
+			throw new IllegalArgumentException("The topicName has not been specified.");
+		}
+	}
+
+
+	private transient ManagedChannel managedChannel = null;
+	private transient TransportChannel channel = null;
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		Publisher.Builder builder = Publisher
+			.newBuilder(ProjectTopicName.of(projectName, topicName))
+			.setCredentialsProvider(serializableCredentialsProvider);
+
+		if (hostAndPort != null) {
+			managedChannel = ManagedChannelBuilder
+				.forTarget(hostAndPort)
+				.usePlaintext(true) // This is 'Ok' because this is ONLY used for testing.
+				.build();
+			channel = GrpcTransportChannel.newBuilder().setManagedChannel(managedChannel).build();
+			builder.setChannelProvider(FixedTransportChannelProvider.create(channel));
+		}
+
+		publisher = builder.build();
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		publisher.shutdown();
```

Comment:
if the `shutdown` throws an exception, the channel may not be closed. so use try/catch warp it.